### PR TITLE
Rewrite partial rowset into full rowset

### PR DESCRIPTION
**Pull request by sevev**
Rewrite partial rowset into full rowset.
Merge partial segment data and read_columns data in memory into a new segment.

### CODE REVIEW

File: be/src/storage/rowset/beta_rowset_writer.cpp [hunk 1]
```diff
@@ -337,8 +337,7 @@ OLAPStatus HorizontalBetaRowsetWriter::flush_chunk(const vectorized::Chunk& chun
     return OLAP_SUCCESS;
 }
 
-OLAPStatus HorizontalBetaRowsetWriter::flush_chunk_with_deletes(const vectorized::Chunk& upserts,
-                                                                const vectorized::Column& deletes) {
+OLAPStatus HorizontalBetaRowsetWriter::de(const vectorized::Chunk& upserts, const vectorized::Column& deletes) {
```
Comment (Linkerist) [inline be/src/storage/rowset/beta_rowset_writer.cpp:340]: de?

**Comment by sevev** [comment]
run starrocks_be_unittest

File: be/src/storage/rowset/segment_rewriter.cpp [hunk 2]
```diff
@@ -2,18 +2,116 @@
...
+    uint64_t file_size;
+    RETURN_IF_ERROR(rblock->size(&file_size));
+
+    if (file_size < 12) {
+        return Status::Corruption(strings::Substitute("Bad segment file $0: file size $1 < 12", src, file_size));
+    }
+
+    SegmentFooterPB footer;
+    size_t footer_length = 0;
+    RETURN_IF_ERROR(parse_segment_footer(&rblock, &footer, &footer_length));
```
Comment (decster) [inline be/src/storage/rowset/segment_rewriter.cpp:79]: How about moving this into parse_segment_footer, then returning a body length(as remaining)

File: be/src/storage/rowset/segment_rewriter.h [hunk 3]
```diff
@@ -7,6 +7,8 @@
 #include <vector>
 
 #include "common/statusor.h"
+#include "gen_cpp/segment.pb.h"
+#include "storage/fs/block_manager.h"
```
Comment (decster) [inline be/src/storage/rowset/segment_rewriter.h:11]: Move to cpp?

File: be/src/storage/rowset/segment_rewriter.cpp [hunk 4]
```diff
@@ -2,18 +2,116 @@
...
+#include "util/slice.h"
 
+namespace starrocks {
 namespace segment_v2 {
 
 SegmentRewriter::SegmentRewriter() {}
 
 SegmentRewriter::~SegmentRewriter() {}
 
+Status parse_segment_footer(std::unique_ptr<fs::ReadableBlock>* rblock, SegmentFooterPB* footer,
```
Comment (Linkerist) [inline be/src/storage/rowset/segment_rewriter.cpp:19]: similar with https://github.com/StarRocks/starrocks/blob/d78926e4ad897c016cbde18a4a2c394d4d5b84fb/be/src/storage/rowset/segment.cpp#L129-L206, can reuse?

File: be/src/storage/rowset/segment_rewriter.cpp [hunk 5]
```diff
@@ -2,18 +2,116 @@
...
+                                    (*rblock)->path(), actual_checksum, checksum));
+    }
+
+    if (!footer->ParseFromArray(buff.data(), buff.size())) {
+        return Status::Corruption(
+                strings::Substitute("Bad segment file $0: failed to parse footer", (*rblock)->path()));
+    }
+
+    return Status::OK();
+}
```
Comment (Linkerist) [inline be/src/storage/rowset/segment_rewriter.cpp:57]: similar with https://github.com/StarRocks/starrocks/blob/d78926e4ad897c016cbde18a4a2c394d4d5b84fb/be/src/storage/rowset/segment.cpp#L129-L206, can reuse?

File: be/src/storage/rowset/segment_rewriter.cpp (hunk 2 above)
Comment (sevev) [reply be/src/storage/rowset/segment_rewriter.cpp]: > How about moving this into parse_segment_footer, then returning a body length(as remaining)

done

File: be/src/storage/rowset/segment_rewriter.h (hunk 3 above)
Comment (sevev) [reply be/src/storage/rowset/segment_rewriter.h]: done

File: be/src/storage/rowset/segment_rewriter.cpp (hunk 5 above)
Comment (sevev) [reply be/src/storage/rowset/segment_rewriter.cpp]: > similar with
> 
> https://github.com/StarRocks/starrocks/blob/d78926e4ad897c016cbde18a4a2c394d4d5b84fb/be/src/storage/rowset/segment.cpp#L129-L206
> 
> , can reuse?

done

**Review by decster — DISMISSED** [review]

File: be/src/storage/rowset/segment.cpp [hunk 6]
```diff
@@ -72,70 +72,15 @@ StatusOr<std::shared_ptr<Segment>> Segment::open(MemTracker* mem_tracker, fs::Bl
...
-        }
-        adapter->set_iterator(std::move(result.value()));
-        return std::move(adapter);
-    } else {
-        return _new_iterator(schema, read_options);
-    }
-}
-
-Status Segment::_parse_footer(size_t* footer_length_hint, SegmentFooterPB* footer) {
+Status Segment::parse_segment_footer(std::unique_ptr<fs::ReadableBlock>* rblock, SegmentFooterPB* footer,
```
Comment (chaoyli) [inline be/src/storage/rowset/segment.cpp:75]: Why use unique_ptr pointer?

File: be/src/storage/rowset/segment.cpp [hunk 7]
```diff
@@ -192,19 +139,82 @@ Status Segment::_parse_footer(size_t* footer_length_hint, SegmentFooterPB* foote
...
+    DCHECK_NOTNULL(read_options.stats);
+    // trying to prune the current segment by segment-level zone map
+    for (const auto& pair : read_options.predicates) {
+        ColumnId column_id = pair.first;
+        if (_column_readers[column_id] == nullptr || !_column_readers[column_id]->has_zone_map()) {
+            continue;
+        }
+        if (!_column_readers[column_id]->segment_zone_map_filter(pair.second)) {
+            read_options.stats->segment_stats_filtered += _column_readers[column_id]->num_rows();
+            return Status::EndOfFile("empty iterator");
```
Comment (chaoyli) [inline be/src/storage/rowset/segment.cpp:189]: Add file name to this log.

File: be/src/storage/rowset/segment_rewriter.cpp [hunk 8]
```diff
@@ -2,8 +2,18 @@
...
+#include "storage/fs/block_manager.h"
+#include "storage/fs/fs_util.h"
+#include "storage/rowset/segment.h"
+#include "storage/rowset/segment_writer.h"
+#include "storage/vectorized/chunk_helper.h"
+#include "util/crc32c.h"
+#include "util/slice.h"
 
+namespace starrocks {
 namespace segment_v2 {
```
Comment (chaoyli) [inline be/src/storage/rowset/segment_rewriter.cpp:17]: segment_v2 namespace is already removed.

File: be/src/storage/rowset/segment.cpp (hunk 6 above)
Comment (sevev) [reply be/src/storage/rowset/segment.cpp]: > Why use unique_ptr pointer?

done

File: be/src/storage/rowset/segment_rewriter.cpp (hunk 8 above)
Comment (sevev) [reply be/src/storage/rowset/segment_rewriter.cpp]: done

**Comment by sevev** [comment]
run all

**Comment by sevev** [comment]
run starrocks_fe_unittest

**Comment by sevev** [comment]
run starrocks_tscancode

**Comment by sevev** [comment]
run starrocks_be_unittest

**Review by decster — DISMISSED** [review]

File: be/src/storage/rowset/segment.cpp [hunk 9]
```diff
@@ -192,19 +138,82 @@ Status Segment::_parse_footer(size_t* footer_length_hint, SegmentFooterPB* foote
...
     }
+
+    if (segment_data_size) {
+        *segment_data_size = file_size - footer_length - 12;
+    }
+
     return Status::OK();
 }
 
+Segment::Segment(const private_type&, fs::BlockManager* blk_mgr, std::string fname, uint32_t segment_id,
```
Comment (Linkerist) [inline be/src/storage/rowset/segment.cpp:160]: if the parameter private_type is useless, maybe should remove this parameter.

File: be/src/storage/rowset/segment_rewriter.cpp [hunk 10]
```diff
@@ -2,21 +2,73 @@
...
-    // TODO: impl
+                                std::vector<std::unique_ptr<vectorized::Column>>& columns, size_t segment_id) {
+    fs::BlockManager* block_mgr = fs::fs_util::block_manager();
+    std::unique_ptr<fs::WritableBlock> wblock;
+    fs::CreateBlockOptions wblock_opts({dest});
+    wblock_opts.mode = Env::CREATE_OR_OPEN_WITH_TRUNCATE;
+    RETURN_IF_ERROR(block_mgr->create_block(wblock_opts, &wblock));
+
+    std::unique_ptr<fs::ReadableBlock> rblock;
+    RETURN_IF_ERROR(block_mgr->open_block(src, &rblock));
```
Comment (Linkerist) [inline be/src/storage/rowset/segment_rewriter.cpp:32]: maybe src_path, dst_path, tablet_schema are better variable name.

File: be/src/storage/rowset/segment_rewriter.cpp [hunk 11]
```diff
@@ -2,21 +2,73 @@
...
+    opts.storage_format_version = config::storage_format_version;
+    SegmentWriter writer(std::move(wblock), segment_id, &tschema, opts);
+    RETURN_IF_ERROR(writer.init(column_ids, false, &footer));
+
+    auto schema = vectorized::ChunkHelper::convert_schema_to_format_v2(tschema, column_ids);
+    auto chunk = vectorized::ChunkHelper::new_chunk(schema, columns[0]->size());
+    for (int i = 0; i < columns.size(); ++i) {
+        vectorized::ColumnPtr& dst_col = chunk->get_column_by_index(i);
+        vectorized::Column* src_col = columns[i].release();
+        dst_col.reset(src_col);
```
Comment (Linkerist) [inline be/src/storage/rowset/segment_rewriter.cpp:63]:
```suggestion
        chunk->get_column_by_index(i).reset(columns[i].release());
```

File: be/src/storage/rowset/segment_writer.cpp [hunk 12]
```diff
@@ -80,6 +80,25 @@ Status SegmentWriter::init(const std::vector<uint32_t>& column_indexes, bool has
...
 
+    // merge partial segment footer
+    // in partial update, key columns and some value columns have been written in partial segment
+    // rewrite partial segment into full segment only need to write other value columns into full segment
+    // merge partial segment footer to avoid loss of metadata
+    if (footer != nullptr) {
+        for (uint32_t ordinal = 0; ordinal < footer->columns().size(); ++ordinal) {
+            const auto& src_column_pb = footer->columns(ordinal);
+            ColumnMetaPB* dst_column_pb = _footer.add_columns();
+            *dst_column_pb = src_column_pb;
```
Comment (Linkerist) [inline be/src/storage/rowset/segment_writer.cpp:91]:
```suggestion
            *_footer.add_columns() = footer->columns(ordinal);
```

File: be/src/storage/rowset/segment_writer.cpp [hunk 13]
```diff
@@ -80,6 +80,25 @@ Status SegmentWriter::init(const std::vector<uint32_t>& column_indexes, bool has
...
+        }
+        if (footer->has_short_key_index_page()) {
+            PagePointerPB* dst = _footer.mutable_short_key_index_page();
+            *dst = footer->short_key_index_page();
```
Comment (Linkerist) [inline be/src/storage/rowset/segment_writer.cpp:95]:
```suggestion
            *_footer.mutable_short_key_index_page() = footer->short_key_index_page();
```

Comment (sevev) [reply be/src/storage/rowset/segment_writer.cpp]: done

File: be/src/storage/rowset/segment_writer.cpp (hunk 12 above)
Comment (sevev) [reply be/src/storage/rowset/segment_writer.cpp]: done

File: be/src/storage/rowset/segment_rewriter.cpp (hunk 11 above)
Comment (sevev) [reply be/src/storage/rowset/segment_rewriter.cpp]: done

File: be/src/storage/rowset/segment_rewriter.cpp (hunk 10 above)
Comment (sevev) [reply be/src/storage/rowset/segment_rewriter.cpp]: done

**Review by decster — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**